File: TextMenu.cc
#include "TextMenu.hh"

#include <cmath>
#include <string>
#include <vector>

#include <SFML/Graphics.hpp>

namespace CastleEscape {

TextMenu::TextMenu(const std::vector<std::string>& options,
		const sf::Font& font) {
	this->options = options;
	this->font = font;
	defaultSpacing = 1;
	canMove = false;
	isFinished = false;
	canPressZ = false;
}

void TextMenu::Update(const sf::Clock& clock, const sf::Input& input) {
	if (!isFinished) {
		selectedSpacing = (float)(defaultSpacing + std::sin(clock.GetElapsedTime() * 5));
		selectedSpacing = 1 + selectedSpacing / options[selectedOption].size();
	} else {
		canMove = false;
		selectedSpacing = defaultSpacing;
		return;
	}

	if (!input.IsKeyDown(sf::Key::Z))
		canPressZ = true;

	if (canPressZ && input.IsKeyDown(sf::Key::Z)) {
		isFinished = true;
		canPressZ = false;
		selectedSpacing = defaultSpacing;
	}

	if (!input.IsKeyDown(sf::Key::Up) && !input.IsKeyDown(sf::Key::Down))
		canMove = true;
	if (!canMove)
		return;
	if (input.IsKeyDown(sf::Key::Up)) {
		selectedOption = selectedOption > 0 ? selectedOption - 1
				: options.size() - 1;
		canMove = false;
	} else if (input.IsKeyDown(sf::Key::Down)) {
		selectedOption = (selectedOption + 1) % options.size();
		canMove = false;
	}
}

void TextMenu::Draw(sf::RenderWindow& window, int x, int y,
		const sf::Color& color) {
	/*
	 for (int i = 0; i < options.Length; i++) {
	 var pos = new Vector2(x, y + i * font.LineSpacing);
	 if (i == selectedOption && !isFinished) {
	 pos.X += 2 * selectedSpacing;
	 font.Spacing = selectedSpacing;
	 } else
	 font.Spacing = defaultSpacing;
	 spriteBatch.DrawString(font, options[i], pos, textColor);
	 }
	 font.Spacing = defaultSpacing;
	 */
	for (uint i = 0; i < options.size(); i++) {
		sf::String item(options[i], font);
		item.SetPosition(x, y + i * font.GetCharacterSize());
		item.SetColor(color);
		if (i == selectedOption)
			item.SetScaleX(selectedSpacing);
		window.Draw(item);
	}
}

} // namespace CastleEscape

File: TextMenu.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SFML;
using SFML.Graphics;
using SFML.Window;

namespace CastleEscape
{
    /// <summary>
    /// Allows the user to select from a list of options.
    /// 
    /// Author: Dennis Honeyman
    /// </summary>
    class TextMenu
    {
        private Font font;
        private string[] options;
        private int selectedOption = 0;
        private float defaultStretch;
        private float selectedStretch;
        private bool canMove;
        private bool isFinished;
        private bool canPressZ;

        public bool IsFinished
        {
            get { return isFinished; }
            set { isFinished = value; }
        }

        public int SelectedOption
        {
            get { return selectedOption; }
        }

        public string[] Options
        {
            get { return options; }
            set { options = value; }
        }

        public TextMenu(Font font, string[] options)
        {
            this.options = options;
            this.font = font;
            defaultStretch = 1;
            canMove = false;
            isFinished = false;
            canPressZ = false;
        }

        public void Update(Clock clock, Input input)
        {
            if (!isFinished)
            {
                selectedStretch = (float)(defaultStretch + Math.Sin(clock.TotalTime.Second * 20));
                selectedStretch = selectedStretch / options[selectedOption].Length;
            }
            else
            {
                canMove = false;
                selectedStretch = defaultStretch;
                return;
            }

            if (state.IsKeyUp(Keys.Z))
                canPressZ = true;

            if (canPressZ && state.IsKeyDown(Keys.Z))
            {
                isFinished = true;
                canPressZ = false;
                selectedStretch = defaultStretch;
            }

            if (state.IsKeyUp(Keys.Up) && state.IsKeyUp(Keys.Down))
                canMove = true;
            if (!canMove)
                return;
            if (state.IsKeyDown(Keys.Up))
            {
                selectedOption = selectedOption > 0 ? selectedOption - 1 : options.Length - 1;
                canMove = false;
            }
            else if (state.IsKeyDown(Keys.Down))
            {
                selectedOption = (selectedOption + 1) % options.Length;
                canMove = false;
            }
        }

        public void Draw(RenderWindow window, int x, int y, Color textColor)
        {
            for (int i = 0; i < options.Length; i++)
            {
				var optionText = new String2D(options[i], font);
                var pos = new Vector2(x, y + i * font.CharacterSize);
                if (i == selectedOption && !isFinished)
                {
                    pos.X += 2 * selectedStretch;
                    optionText.Scale = new Vector2(selectedStretch, 0);
                }
				optionText.Position = pos;
				optionText.Color = textColor;
				window.Draw(optionText);
            }
        }
    }
}
